lib/pos_epoch_complete_hook.go

```diff
@@ -266,13 +273,15 @@ func (bav *UtxoView) runEpochCompleteEpochRollover(epochNumber uint64, blockHeig
 	return nil
 }
 
-func (bav *UtxoView) generateAndSnapshotValidatorSet(epochNumber uint64) ([]*ValidatorEntry, error) {
+func (bav *UtxoView) generateAndSnapshotValidatorSet(epochNumber uint64, maxValidatorSet []*ValidatorEntry,
+) []*ValidatorEntry {
 	// Snapshot the current top n active validators as the validator set.
-	validatorSet, err := bav.GetTopActiveValidatorsByStakeAmount(
-		bav.GetCurrentGlobalParamsEntry().ValidatorSetMaxNumValidators,
-	)
-	if err != nil {
-		return nil, errors.Wrapf(err, "generateAndSnapshotValidatorSet: error retrieving top ValidatorEntries: ")
+	validatorSetMaxNumValidators := bav.GetCurrentGlobalParamsEntry().ValidatorSetMaxNumValidators
+	var validatorSet []*ValidatorEntry
+	if len(maxValidatorSet) > int(validatorSetMaxNumValidators) {
+		validatorSet = maxValidatorSet[:validatorSetMaxNumValidators]
+	} else {
+		validatorSet = maxValidatorSet[:]
```

Comment:
Why filter or create a copy? It's safe to assume that the validator set passed in is already limited to the max size
```suggestion
```